beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java

```diff
@@ -302,11 +317,16 @@ void stop() {
   }
 
   void stop(boolean nack) {
+    nackAllIncoming = nack;
     stopped.set(true);
     if (nack) {
       List<Pair<StreamElement, UnifiedContext>> drop = new ArrayList<>();
       queue.drainTo(drop);
       drop.stream().map(Pair::getSecond).filter(Objects::nonNull).forEach(UnifiedContext::nack);
+      Optional.ofNullable(lastWrittenContext).ifPresent(UnifiedContext::nack);
+    }
+    if (getWatermark() < Watermarks.MAX_WATERMARK) {
+      ExceptionUtils.ignoringInterrupted(() -> cancelledLatch.await(1, TimeUnit.SECONDS));
```

Comment:
what are the consequences of latch timing out?

Reply:
Hm, interesting question. It might result in element not being correctly `nack`ed, which would in most implementations probably result in that this element would be retried after some timeout, but will be probably late. The other option of not setting a timeout for the latch might result (probably under the same circumstances as the the timeout in the other case) would result in pipeline being stuck, which seems to be worse case IMO.